### ISSUE TITLE
List-backups broken in between v0.15.0 and v0.16.0

### ISSUE THREAD

**Issue by FHRNet** [thelastpickle/cassandra-medusa#657]
[Project board link](https://github.com/orgs/k8ssandra/projects/8/views/1?pane=issue&itemId=40011579)

Hi, it seems that the "list-backups" command is broken on v0.16.0 and v0.16.1, with the last working release being v0.15.0.

Storage provider used - `s3_compatible` (Minio)

Both of the following outputs were captured on the same node. The correct listing contains two backups named "2023_09_26" and "2023_09_27", however both are completely absent on >=v0.16.0 from the output.

Debug log with v0.16.0:
```
$ sudo pip3 install --upgrade cassandra-medusa[S3]==0.16.0
$ medusa --version
0.16.0
$ medusa -v list-backups
[2023-09-28 01:54:16,376] DEBUG: Loading configuration from /etc/medusa/medusa.ini
[2023-09-28 01:54:16,377] INFO: Resolving ip address
[2023-09-28 01:54:16,377] INFO: ip address to resolve 192.0.2.1
[2023-09-28 01:54:16,377] DEBUG: Not resolving 192.0.2.1 as requested
[2023-09-28 01:54:16,377] DEBUG: Loading storage_provider: s3_compatible
[2023-09-28 01:54:16,381] DEBUG: Setting AWS credentials file to /etc/medusa/medusa-minio-credentials
[2023-09-28 01:54:16,384] INFO: Using credentials CensoredCredentials(access_key_id=T..d, secret_access_key=*****, region=de1)
[2023-09-28 01:54:16,384] INFO: Using S3 URL https://minio.repository.example.com:443
[2023-09-28 01:54:16,384] DEBUG: Connecting to S3
[2023-09-28 01:54:16,411] DEBUG: [Storage] Listing objects in index/backup_index
[2023-09-28 01:54:16,411] DEBUG: Using selector: GeventSelector
[2023-09-28 01:54:17,716] DEBUG: Blob 192.0.2.2/2023_08_30/meta/schema.cql was not found in cache.
[2023-09-28 01:54:17,716] DEBUG: [Storage] Getting object 192.0.2.2/2023_08_30/meta/schema.cql
[2023-09-28 01:54:17,906] DEBUG: Disconnecting from S3...
```

Debug log with v0.15.0:
```
$ sudo pip3 install --upgrade cassandra-medusa[S3]==0.15.0
$ medusa --version
0.15.0
$ medusa -v list-backups
[2023-09-28 01:53:39,906] DEBUG: Loading configuration from /etc/medusa/medusa.ini
[2023-09-28 01:53:39,906] INFO: Resolving ip address
[2023-09-28 01:53:39,906] INFO: ip address to resolve 192.0.2.1
[2023-09-28 01:53:39,906] DEBUG: Not resolving 192.0.2.1 as requested
[2023-09-28 01:53:39,907] DEBUG: Loading storage_provider: s3_compatible
[2023-09-28 01:53:39,907] DEBUG: Changing event name from creating-client-class.iot-data to creating-client-class.iot-data-plane
[2023-09-28 01:53:39,908] DEBUG: Changing event name from before-call.apigateway to before-call.api-gateway
[2023-09-28 01:53:39,909] DEBUG: Changing event name from request-created.machinelearning.Predict to request-created.machine-learning.Predict
[2023-09-28 01:53:39,909] DEBUG: Changing event name from before-parameter-build.autoscaling.CreateLaunchConfiguration to before-parameter-build.auto-scaling.CreateLaunchConfiguration
[2023-09-28 01:53:39,909] DEBUG: Changing event name from before-parameter-build.route53 to before-parameter-build.route-53
[2023-09-28 01:53:39,910] DEBUG: Changing event name from request-created.cloudsearchdomain.Search to request-created.cloudsearch-domain.Search
[2023-09-28 01:53:39,910] DEBUG: Changing event name from docs.*.autoscaling.CreateLaunchConfiguration.complete-section to docs.*.auto-scaling.CreateLaunchConfiguration.complete-section
[2023-09-28 01:53:39,911] DEBUG: Changing event name from before-parameter-build.logs.CreateExportTask to before-parameter-build.cloudwatch-logs.CreateExportTask
[2023-09-28 01:53:39,911] DEBUG: Changing event name from docs.*.logs.CreateExportTask.complete-section to docs.*.cloudwatch-logs.CreateExportTask.complete-section
[2023-09-28 01:53:39,911] DEBUG: Changing event name from before-parameter-build.cloudsearchdomain.Search to before-parameter-build.cloudsearch-domain.Search
[2023-09-28 01:53:39,911] DEBUG: Changing event name from docs.*.cloudsearchdomain.Search.complete-section to docs.*.cloudsearch-domain.Search.complete-section
[2023-09-28 01:53:39,912] DEBUG: Setting config variable for region to 'de1'
[2023-09-28 01:53:39,912] DEBUG: Setting AWS credentials file to /etc/medusa/medusa-minio-credentials
[2023-09-28 01:53:39,912] DEBUG: Setting config variable for credentials_file to '/etc/medusa/medusa-minio-credentials'
[2023-09-28 01:53:39,913] DEBUG: IMDS ENDPOINT: http://169.254.169.254/
[2023-09-28 01:53:39,914] DEBUG: Looking for credentials via: env
[2023-09-28 01:53:39,914] DEBUG: Looking for credentials via: assume-role
[2023-09-28 01:53:39,914] DEBUG: Looking for credentials via: assume-role-with-web-identity
[2023-09-28 01:53:39,914] DEBUG: Looking for credentials via: sso
[2023-09-28 01:53:39,914] DEBUG: Looking for credentials via: shared-credentials-file
[2023-09-28 01:53:39,914] INFO: Found credentials in shared credentials file: /etc/medusa/medusa-minio-credentials
[2023-09-28 01:53:40,649] DEBUG: [Storage] Listing objects in index/backup_index
[2023-09-28 01:53:41,716] DEBUG: Blob 192.0.2.2/2023_08_30/meta/schema.cql was not found in cache.
[2023-09-28 01:53:41,716] DEBUG: [Storage] Getting object 192.0.2.2/2023_08_30/meta/schema.cql
[2023-09-28 01:53:42,451] DEBUG: [Storage] Reading blob index/backup_index/2023_09_26/tokenmap_192.0.2.1.json...
[2023-09-28 01:53:42,451] DEBUG: [Storage] Reading blob index/backup_index/2023_09_26/tokenmap_192.0.2.1.json...
2023_09_26 (started: 2023-09-26 19:16:09, finished: 2023-09-27 00:34:29)
[2023-09-28 01:53:43,012] DEBUG: [Storage] Reading blob index/backup_index/2023_09_27/tokenmap_192.0.2.1.json...
[2023-09-28 01:53:43,012] DEBUG: [Storage] Reading blob index/backup_index/2023_09_27/tokenmap_192.0.2.1.json...
2023_09_27 (started: 2023-09-27 19:16:08, finished: 2023-09-27 21:19:04)
```

The listed backups also cannot be found using `medusa status --backup-name` on v0.16.0, an `ERROR: No such backup` message is emitted. These backups were taken on the v0.16.1 release and as per the output, succeeded.

In other words, it seems it might be an issue specifically related to reading/listing the state

**Comment by rzvoncek** [comment]
Hi @FHRNet, 

This line in the 0.16.0 trace looks suspicious:
```
[2023-09-28 01:54:16,384] INFO: Using S3 URL https://minio.repository.example.com:443
```
It's probably not where your minio is running.

Could you please check (and tell me) what do values do you have for the `host` and `port` items in the `storage` section in your `/etc/medusa/medusa.ini` ?

**Comment by FHRNet** [comment]
Hi @rzvoncek, I have indeed redacted the IP addresses and the minio repository URL, those values are fine and the configs did not change in between the version change.

I've done some additional testing and it seems like the root cause is the s3_compatible driver not returning all the files on v0.16.0. I edited the `AbstractStorage.list_objects` method to see what we're getting and the returned list is truncated to exactly 1000 objects, meanwhile on v0.15.0 we get 1368 entries back.

```
    @retry(stop_max_attempt_number=7, wait_exponential_multiplier=10000, wait_exponential_max=120000)
    def list_objects(self, path=None):
        # List objects in the bucket/container that have the corresponding prefix (emtpy means all objects)
        logging.debug("[Storage] Listing objects in {}".format(path if path is not None else 'everywhere'))

        objects = self.list_blobs(prefix=path)

+        for object in objects:
+            print(object.name)

        objects = list(filter(lambda blob: blob.size > 0, objects))

        return objects
```

**Comment by rzvoncek** [comment]
Hi @FHRNet , thanks for the pointer. There is/was indeed a problem with pagination. In 0.16 we switched how we talked to storage, and out of the 3 new libraries we use, only one was paginating transparently. 

I've now added the pagination to the other two in #658, including some tests that reproduced the "listing only 1000 object when I expected more" issue.

If you have a chance to test this off the branch, please do. Otherwise please wait for a new minor of 0.16, which I believe we'll release shortly.

**Comment by rzvoncek** [comment]
Hi @FHRNet , v0.16.2 is now available at PiPy: https://pypi.org/project/cassandra-medusa/#history

**Comment by FHRNet** [comment]
Hi @rzvoncek I had a chance to test 0.16.2, I can confirm the issue is now indeed fixed and the missing backups are present.

Thank you for the quick fix!
